File: src/pipeline_tracker.rs
use chrono::{Utc, DateTime};
use crate::error::TrackerError;
use crate::models::{DiscordMessage, StepInfo, StepStatus};
use crate::storage::{MessageStorage, PipelineState};
use crate::message_builder::{build_init_embed, build_step_update_embed, build_completion_embed};
use crate::discord_api::DiscordApi;
use crate::validation::validate_step_number;

/// Main pipeline tracker that orchestrates Discord notifications
pub struct PipelineTracker {
    api: DiscordApi,
    storage: MessageStorage,
    message_id: Option<String>,
    steps: Vec<StepInfo>,
    pr_info: Option<PrInfo>,
    pipeline_started_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone)]
struct PrInfo {
    number: String,
    title: String,
    author: String,
    repository: String,
    branch: String,
}

impl PipelineTracker {
    /// Creates a new pipeline tracker
    pub fn new(bot_token: &str, channel_id: &str) -> Result<Self, TrackerError> {
        let api = DiscordApi::new(bot_token, channel_id)?;
        let storage = MessageStorage::new()?;
        
        Ok(Self {
            api,
            storage,
            message_id: None,
            steps: Vec::new(),
            pr_info: None,
            pipeline_started_at: None,
        })
    }
    
    /// Initializes the pipeline tracking
    pub async fn init_pipeline(
        &mut self,
        pr_number: &str,
        pr_title: &str,
        author: &str,
        repository: &str,
        branch: &str,
    ) -> Result<(), TrackerError> {
        self.pr_info = Some(PrInfo {
            number: pr_number.to_string(),
            title: pr_title.to_string(),
            author: author.to_string(),
            repository: repository.to_string(),
            branch: branch.to_string(),
        });
        
        self.pipeline_started_at = Some(Utc::now());
        
        let embed = build_init_embed(pr_number, pr_title, author, repository, branch);
        let message = DiscordMessage {
            content: None,
            embeds: vec![embed],
        };
        
        let message_id = self.api.send_message(&message).await?;
        self.message_id = Some(message_id.clone());
        
        // Save state
        let state = PipelineState {
            message_id,
            pr_number: pr_number.parse().unwrap_or(0),
            pr_title: pr_title.to_string(),
            author: author.to_string(),
            repository: repository.to_string(),
            branch: branch.to_string(),
            steps: self.steps.clone(),
            pipeline_started_at: self.pipeline_started_at.unwrap(),
        };
        self.storage.save_pipeline_state(&state).await?;
        
        Ok(())
    }
    
    /// Updates a step in the pipeline
    pub async fn update_step(
        &mut self,
        step_number: u32,
        total_steps: u32,
        step_name: &str,
        status: &str,
        additional_info: &[(String, String)],
    ) -> Result<(), TrackerError> {
        validate_step_number(step_number, total_steps)?;
        
        let step_status = StepStatus::from_str(status)
            .map_err(|e| TrackerError::InvalidStatus(e))?;
        
        // Find or create step
        let step_index = self.steps.iter().position(|s| s.number == step_number);
        if let Some(index) = step_index {
            // Update existing step
            let step = &mut self.steps[index];
            step.name = step_name.to_string();
            step.status = step_status.clone();
            step.additional_info = additional_info.to_vec();
        } else {
            // Create new step
            let new_step = StepInfo::new(
                step_number,
                step_name.to_string(),
                step_status.clone(),
                additional_info.to_vec(),
            );
            self.steps.push(new_step);
        }
        
        // Mark step as completed if it's finished
        if step_status == StepStatus::Success || step_status == StepStatus::Failed {
            if let Some(index) = self.steps.iter().position(|s| s.number == step_number) {
                self.steps[index].mark_completed();
            }
        }
        
        // Update Discord message
        if let Some(pr_info) = &self.pr_info {
            let embed = build_step_update_embed(
                &pr_info.number,
                &pr_info.title,
                &self.steps,
                step_number,
                total_steps,
            );
            
            let message = DiscordMessage {
                content: None,
                embeds: vec![embed],
            };
            
            if let Some(message_id) = &self.message_id {
                self.api.update_message(message_id, &message).await?;
            }
        }
        
        // Save state
        if let Some(pr_info) = &self.pr_info {
            let state = PipelineState {
                message_id: self.message_id.clone().unwrap_or_default(),
                pr_number: pr_info.number.parse().unwrap_or(0),
                pr_title: pr_info.title.clone(),
                author: pr_info.author.clone(),
                repository: pr_info.repository.clone(),
                branch: pr_info.branch.clone(),
                steps: self.steps.clone(),
                pipeline_started_at: self.pipeline_started_at.unwrap_or_else(Utc::now),
            };
            self.storage.save_pipeline_state(&state).await?;
        }
        
        Ok(())
    }
    
    /// Completes the pipeline
    pub async fn complete_pipeline(&mut self) -> Result<(), TrackerError> {
        if let (Some(pr_info), Some(start_time)) = (&self.pr_info, self.pipeline_started_at) {
            let total_steps = self.steps.len() as u32;
            let embed = build_completion_embed(
                &pr_info.number,
                &pr_info.title,
                &self.steps,
                total_steps,
                start_time,
            );
            
            let message = DiscordMessage {
                content: None,
                embeds: vec![embed],
            };
            
            if let Some(message_id) = &self.message_id {
                self.api.update_message(message_id, &message).await?;
            }
        }
        
        // Clear state
        self.storage.clear_pipeline_state().await?;
        
        Ok(())
    }
} 

File: src/pipelineTracker.ts
import { DiscordApi } from './discordApi';
import { TrackerError } from './error';
import { buildCompletionEmbed, buildInitEmbed, buildStepUpdateEmbed } from './messageBuilder';
import {
  type DiscordMessage,
  type PipelineState,
  type PrInfo,
  type StepInfo,
  StepInfoManager,
  StepStatus,
  StepStatusHelper,
} from './models';

// Internal pipeline state interface (matches file storage format)
export interface InternalPipelineState {
	messageId: string;
	prNumber: number;
	prTitle: string;
	author: string;
	repository: string;
	branch: string;
	steps: StepInfo[];
	pipelineStartedAt: Date;
}

// Storage interface for dependency injection
export interface Storage {
	savePipelineState(state: InternalPipelineState): Promise<void>;
	clearPipelineState(): Promise<void>;
	loadPipelineState(): Promise<InternalPipelineState | null>;
}

// In-memory storage implementation (for testing and simple use cases)
export class InMemoryStorage implements Storage {
  private state: InternalPipelineState | null = null;

  async savePipelineState(state: InternalPipelineState): Promise<void> {
    this.state = state;
  }

  async clearPipelineState(): Promise<void> {
    this.state = null;
  }

  async loadPipelineState(): Promise<InternalPipelineState | null> {
    return this.state;
  }
}

/// Main pipeline tracker that orchestrates Discord notifications
export class PipelineTracker {
  private api: DiscordApi;
  private storage: Storage;
  private messageId: string | undefined;
  private steps: StepInfo[];
  private prInfo: PrInfo | undefined;
  private pipelineStartedAt: Date | undefined;

  constructor(botToken: string, channelId: string, storage?: Storage) {
    this.api = new DiscordApi(botToken, channelId);
    this.storage = storage || new InMemoryStorage();
    this.messageId = undefined;
    this.steps = [];
    this.prInfo = undefined;
    this.pipelineStartedAt = undefined;
  }

  /// Initializes the pipeline tracking
  async initPipeline(
    prNumber: string,
    prTitle: string,
    author: string,
    repository: string,
    branch: string
  ): Promise<void> {
    this.prInfo = {
      number: prNumber,
      title: prTitle,
      author: author,
      repository: repository,
      branch: branch,
    };

    this.pipelineStartedAt = new Date();

    const embed = buildInitEmbed(prNumber, prTitle, author, repository, branch);
    const message: DiscordMessage = {
      content: '',
      embeds: [embed],
    };

    const messageId = await this.api.sendMessage(message);
    this.messageId = messageId;

    // Save state
    const state: InternalPipelineState = {
    	messageId: messageId,
    	prNumber: parseInt(prNumber, 10) || 0,
    	prTitle: prTitle,
    	author: author,
    	repository: repository,
    	branch: branch,
    	steps: this.steps,
    	pipelineStartedAt: this.pipelineStartedAt,
    };
    await this.storage.savePipelineState(state);
  }

  /// Updates a step in the pipeline
  async updateStep(
    stepNumber: number,
    totalSteps: number,
    stepName: string,
    status: string,
    additionalInfo: [string, string][]
  ): Promise<void> {
    if (stepNumber <= 0 || totalSteps <= 0 || stepNumber > totalSteps) {
      throw TrackerError.invalidStepNumber(stepNumber);
    }

    const stepStatus = StepStatusHelper.fromStr(status) as StepStatus;
    if (typeof stepStatus !== 'string') {
      throw TrackerError.invalidStatus(status);
    }

    // Find or create step
    let step = this.steps.find((s) => s.number === stepNumber);
    if (step) {
      // Update existing step
      step.name = stepName;
      step.status = stepStatus;
      step.additionalInfo = additionalInfo;
    } else {
      // Create new step
      step = StepInfoManager.new(stepNumber, stepName, stepStatus, additionalInfo);
      this.steps.push(step);
    }

    // Mark step as completed if it's finished
    if (
      stepStatus === StepStatus.Success ||
      stepStatus === StepStatus.Failed ||
      stepStatus === StepStatus.Skipped
    ) {
      StepInfoManager.markCompleted(step);
    }

    // Update Discord message
    if (this.prInfo && this.pipelineStartedAt) {
      const embed = buildStepUpdateEmbed(
        this.prInfo.number,
        this.prInfo.title,
        this.steps,
        stepNumber,
        totalSteps
      );

      const message: DiscordMessage = {
        content: '',
        embeds: [embed],
      };

      if (this.messageId) {
        await this.api.updateMessage(this.messageId, message);
      }
    }

    // Save state
    if (this.prInfo && this.pipelineStartedAt) {
      const state: PipelineState = {
        messageId: this.messageId || '',
        prNumber: parseInt(this.prInfo.number, 10) || 0,
        prTitle: this.prInfo.title,
        author: this.prInfo.author,
        repository: this.prInfo.repository,
        branch: this.prInfo.branch,
        steps: this.steps,
        pipelineStartedAt: this.pipelineStartedAt,
      };
      await this.storage.savePipelineState(state);
    }
  }

  /// Completes the pipeline
  async completePipeline(): Promise<void> {
    // Load state from storage first (critical for GitHub Actions)
    await this.loadState();

    if (this.prInfo && this.pipelineStartedAt) {
      const totalSteps = this.steps.length > 0 ? this.steps.length : 1;
      const embed = buildCompletionEmbed(
        this.prInfo.number,
        this.prInfo.title,
        this.steps,
        totalSteps,
        this.pipelineStartedAt
      );

      const message: DiscordMessage = {
        content: '',
        embeds: [embed],
      };

      if (this.messageId) {
        await this.api.updateMessage(this.messageId, message);
      }
    }

    // Clear state
    await this.storage.clearPipelineState();
  }

  /// Loads pipeline state from storage
  async loadState(): Promise<void> {
    const state = await this.storage.loadPipelineState();
    if (state) {
      this.messageId = state.messageId;
      this.steps = state.steps;
      this.prInfo = {
        number: state.prNumber.toString(),
        title: state.prTitle,
        author: state.author,
        repository: state.repository,
        branch: state.branch,
      };
      // Convert string back to Date object when loading from JSON
      this.pipelineStartedAt = new Date(state.pipelineStartedAt);
    }
  }
}
